# Tree topology inference from multiple sequence alignments using Deep Learning

This repository contains R and Python scripts that were used in the project ["Accurate inference of tree topologies from multiple sequence alignments using deep learning"](https://www.biorxiv.org/content/10.1101/559054v1) 

R scripts can be found in [INDELible directory](https://github.com/SchriderLab/Tree_learning/tree/master/INDELible). They will generate various control files for [INDElible](http://abacus.gene.ucl.ac.uk/software/indelible/) program that simulates MSAs under given tree topology, branch lengths and different substitution as well as indel model parameters.  

Required CRAN R packages:  
[phangorn](https://cran.r-project.org/web/packages/phangorn/index.html)   
[MCMCpack](https://cran.r-project.org/web/packages/MCMCpack/index.html)  
[dplyr](https://cran.r-project.org/web/packages/dplyr/index.html)  
[scales](https://cran.r-project.org/web/packages/scales/index.html)  

1) [indelible_controlgen_INDEL001.R](https://github.com/SchriderLab/Tree_learning/blob/master/INDELible/indelible_controlgen_INDEL001.R) and [indelible_controlgen_NOINDEL.R](https://github.com/SchriderLab/Tree_learning/blob/master/INDELible/indelible_controlgen_NOINDEL.R)    
These scripts generate control files for MSA simulation with (INDEL001) and without (NOINDEL) indels/gaps. The control files will be stored in three directories (topo1, topo2 and topo3) that correspond to three topologies. These scripts are used to generate MSAs for generating TRAINING, VALIDATION and TEST data sets. 

2) [indelible_controlgen_REGIONS_INDEL001.R](https://github.com/SchriderLab/Tree_learning/blob/master/INDELible/indelible_controlgen_REGIONS_INDEL001.R) and [indelible_controlgen_REGIONS_NOINDEL0.R](https://github.com/SchriderLab/Tree_learning/blob/master/INDELible/indelible_controlgen_REGIONS_NOINDEL0.R)  
These scripts generate control files for MSA simulation with (INDEL001) and without (NOINDEL) indels/gaps. The scripts will generate FA, FE, LONG, LONGOUT, SHORT, SHORTINT, SHORTOUT directories each with topo1, topo2 and topo3 subdirectories. These correspond to heterogeneous branch length regions, namely Extended Farris zone (FA), Extended Felsenstein zone (FE), Long branches (LONG), Single long branch (LONGOUT), Short branches (SHORT), Short internal branch (SHORTINT) and Single short branch (SHORTOUT). These MSAs were used to test performance of different tree inference methods.      

3) [indelible_controlgen_REGIONS_INDEL001_FATFAFE.R](https://github.com/SchriderLab/Tree_learning/blob/master/INDELible/indelible_controlgen_REGIONS_INDEL001_FATFAFE.R) and [indelible_controlgen_REGIONS_NOINDEL_FATFAFE.R](https://github.com/SchriderLab/Tree_learning/blob/master/INDELible/indelible_controlgen_REGIONS_NOINDEL_FATFAFE.R)  
These scripts generate control files for MSA simulation with (INDEL001) and without (NOINDEL) indels/gaps. The scripts will generate FA, FAT, FE, and SHORTINT directories each with topo1, topo2 and topo3 subdirectories. These correspond to heterogeneous branch length regions, namely Farris zone (FA), Felsenstein zone (FE), ["Twisted" Farris zone](https://www.sciencedirect.com/science/article/pii/S1055790315002316?via%3Dihub) (FAT) and Short internal branch (SHORTINT). These MSAs were used to test performance of different tree inference methods.

4) [indelible_controlgen_INDEL001_WARNOW.R](https://github.com/SchriderLab/Tree_learning/blob/master/INDELible/indelible_controlgen_INDEL001_WARNOW.R)   
This script generates control files for MSA simulation with no substitutions, only indels. This is the scenario under which maximum likelihood (ML) tree inference has been shown to be statistically inconsistent ([Warnow, 2012](http://currents.plos.org/treeoflife/index.html%3Fp=1609.html)). These MSAs were used to test performance of different tree inference methods.  

Python scripts can be found in [KERAS directory](https://github.com/SchriderLab/Tree_learning/tree/master/KERAS). They are used for building, training, validating and testing Convolutional Neuronal Networks (CNNs). 

1) [keras_CNN_TOPO.py](https://github.com/SchriderLab/Tree_learning/blob/master/KERAS/keras_CNN_TOPO.py) 
This script builds, trainins, validates and tests CNN. As an input it takes TRAINING, VALIDATION and TESTING MSAs generated by INDELible saved in .npy array.    




